apollo-spaceport/src/server.rs

```diff
@@ -0,0 +1,330 @@
+            interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+            interval.tick().await;
+            loop {
+                tokio::select! {
+                    biased;
+                    mopt = rx.recv() => {
+                        tracing::debug!("spaceport triggered");
+                        match mopt {
+                            Some(_msg) => {
+                                for result in extract_graph_usage(&client, task_graph_usage.clone()).await {
+                                    match result {
+                                        Ok(v) => tracing::debug!("Report submission succeeded: {:?}", v),
+                                        Err(e) => tracing::error!("Report submission failed: {}", e),
+                                    }
+                                }
+                            },
+                            None => break
+                        }
+                    },
+                    _ = interval.tick() => {
+                        tracing::debug!("spaceport ticked");
+                        task_total.store(0, Ordering::SeqCst);
+                        for result in extract_graph_usage(&client, task_graph_usage.clone()).await {
+                            match result {
+                                Ok(v) => tracing::debug!("Report submission succeeded: {:?}", v),
+                                Err(e) => tracing::error!("Report submission failed: {}", e),
+                            }
+                        }
+                    }
+                };
+            }
+        });
+        Self {
+            addr,
+            graph_usage,
+            tx,
+            total,
+        }
+    }
+
+    /// Start serving requests.
+    pub async fn serve(self) -> Result<(), Error> {
+        let addr = self.addr;
+        Server::builder()
+            .add_service(ReporterServer::new(self))
+            .serve(addr)
+            .await
+    }
+
+    async fn submit_report(
+        client: &Client,
+        key: String,
+        report: Report,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        tracing::debug!("submitting report: {:?}", report);
+        // Protobuf encode message
+        let mut content = BytesMut::new();
+        report
+            .encode(&mut content)
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        // Create a gzip encoder
+        let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+        // Write our content to our encoder
+        encoder
+            .write_all(&content)
+            .map_err(|e| Status::internal(e.to_string()))?;
+        // Finish encoding and retrieve content
+        let compressed_content = encoder
+            .finish()
+            .map_err(|e| Status::internal(e.to_string()))?;
+        let mut backoff = Duration::from_millis(0);
+        let ingress = match std::env::var("APOLLO_INGRESS") {
```

Comment:
There's no existing environment variable for this and I'm on-board choosing a different approach to how this used to be configured (`endpointUrl`, which actually was just like "endpoint prefix" because it was always (surprisingly) [suffixed with `/api/ingress/traces`](https://github.com/apollographql/apollo-server/blob/16adb42ef5427d6b311fd1462667451d33bdf99c/packages/apollo-server-core/src/plugin/usageReporting/plugin.ts#L299-L301).

Might we call this something more specific though `APOLLO_INGRESS_URL` though? `APOLLO_USAGE_REPORTING_INGRESS_URL`?

(I think the only case this is really needed is for running against our staging stuff though and never for users though.)

Reply:
It's useful for testing as well. I slightly prefer your second suggestion. I'll go with that.

Reply:
done: d6fb85ecc376a0c4ff60ebe20f5cf657d64a5659